janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java

```diff
@@ -241,7 +262,30 @@ public static boolean isEmpty(Condition<?> condition) {
     private static <E extends JanusGraphElement> Or<E> addConstraint(final RelationType type, OrJanusPredicate predicate, List<Object> values, Or<E> or, StandardJanusGraphTx tx) {
         for (int i = 0 ; i < values.size(); i++) {
             final JanusGraphPredicate janusGraphPredicate = predicate.get(i);
-            if (janusGraphPredicate instanceof AndJanusPredicate) {
+            if (janusGraphPredicate instanceof Contain) {
+                //Rewrite contains conditions
+                final Collection childValues = (Collection) values.get(i);
+                if (janusGraphPredicate == Contain.NOT_IN) {
+                    if (childValues.size() == 1) {
+                        addConstraint(type, Cmp.NOT_EQUAL, childValues.iterator().next(), or, tx);
+                    }
+                    // Don't need to handle the case where childValues is empty, because it defaults to
+                    // an or(and()) is added, which is a tautology
```

Comment:
I understand they are semantically the same, but will there be a performance difference (even if very little)?

Reply:
There may be a little difference in performance, but the empty `and` actually serves a functional purpose. Otherwise the surrounding `or` can be empty which results in `or()` being evaluated as `false` instead of `or(and())` which is `true`.